distributed/scheduler.py

```diff
@@ -2328,6 +2335,44 @@ def decide_worker(self, ts: TaskState) -> WorkerState:
             ts.state = "no-worker"
             return ws
 
+        total_nthreads: Py_ssize_t
+        if valid_workers is None:
+            total_nthreads = self._total_nthreads
+        else:
+            total_nthreads = 0
+            for ws in valid_workers:
+                total_nthreads += ws._nthreads
+
+        group_tasks_per_thread: double = (
+            (len(group) / total_nthreads) if total_nthreads > 0 else 0
+        )
+        if group_tasks_per_thread > 2 and sum(map(len, group._dependencies)) < 5:
+            # Group is larger than cluster with very few dependencies; minimize future data transfers.
+            ws = group._last_worker
+            if not (ws and valid_workers is not None and ws not in valid_workers):
+                if (
+                    ws
+                    and ws._occupancy / ws._nthreads / self.get_task_duration(ts)
+                    < group_tasks_per_thread
+                ):
+                    # Schedule sequential tasks onto the same worker until it's filled up.
+                    # Assumes `decide_worker` is being called in priority order.
+                    return ws
+
+                # Pick a new worker for the next few tasks, considering all possible workers
+                worker_pool = (
+                    valid_workers
+                    if valid_workers is not None
+                    else (self._idle_dv or self._workers_dv).values()
+                )
+                ws = min(
+                    worker_pool,
+                    key=partial(self.worker_objective, ts),
+                    default=None,
```

Comment:
This introduces an O(n^2) dependence on the number of workers (unless idle kicks in).  Maybe a concern?